metricflow/dataflow/builder/source_node.py

```diff
@@ -21,19 +29,68 @@ def create_from_data_sets(self, data_sets: Sequence[DataSourceDataSet]) -> Seque
         source_nodes: List[BaseOutput[DataSourceDataSet]] = []
         for data_set in data_sets:
             read_node = ReadSqlSourceNode[DataSourceDataSet](data_set)
-            aggregation_time_dimensions = self._semantic_model.data_source_semantics.get_aggregation_time_dimensions(
+            agg_time_dim_to_measures = (
+                self._semantic_model.data_source_semantics.get_aggregation_time_dimensions_with_measures(
+                    data_set.data_source_reference
+                )
+            )
+            non_additive_measures = self._semantic_model.data_source_semantics.get_non_additive_measures(
                 data_set.data_source_reference
             )
+            measure_name_to_spec = {m.spec.element_name: m.spec for m in data_set.instance_set.measure_instances}
+            instance_set_with_no_measures = data_set.instance_set.transform(RemoveMeasures())
+
             # Dimension sources may not have any measures -> no aggregation time dimensions.
-            if len(aggregation_time_dimensions) == 0:
+            if len(agg_time_dim_to_measures) == 0:
                 source_nodes.append(read_node)
             else:
-                for time_dimension_reference in aggregation_time_dimensions:
-                    source_nodes.append(
-                        MetricTimeDimensionTransformNode[DataSourceDataSet](
+                for time_dimension_reference, measures in agg_time_dim_to_measures.items():
+                    measures_for_data_source = set(measures)
+
+                    # Build all source nodes from non-additive measures
+                    for non_additive_measure in non_additive_measures:
+                        if non_additive_measure not in measures_for_data_source:
```

Comment:
There's a lot of cross-checking gymnastics in this PR. If we make the non-additive attributes part of the MeasureSpec instead of keeping them separated in DataSourceSemantics, does that simplify the other things?

I think it should work to do that although it is a larger refactor underneath this to update the various direct calls to the MeasureSpec initializer.